core/src/consensus.rs

```diff
@@ -606,6 +703,203 @@ impl Tower {
+        slot_history: &SlotHistory,
+    ) -> Result<Self> {
+        info!(
+            "adjusting lockouts (after replay up to {}): {:?}",
+            replayed_root_slot,
+            self.voted_slots()
+        );
+
+        assert_eq!(slot_history.check(replayed_root_slot), Check::Found);
+        // reconcile_blockstore_roots_with_tower() should already have aligned these.
+        assert!(
+            self.root().is_none() || self.root().unwrap() <= replayed_root_slot,
+            format!(
+                "tower root: {:?} >= replayed root slot: {}",
+                self.root().unwrap(),
+                replayed_root_slot
+            )
+        );
+        assert!(
+            self.last_vote == Vote::default() && self.lockouts.votes.is_empty()
+                || self.last_vote != Vote::default() && !self.lockouts.votes.is_empty(),
+            format!(
+                "last vote: {:?} lockouts.votes: {:?}",
+                self.last_vote, self.lockouts.votes
+            )
+        );
+
+        // return immediately if votes are empty...
+        if self.lockouts.votes.is_empty() {
+            assert_eq!(self.root(), None);
+            return Ok(self);
+        }
+
+        let last_voted_slot = self.last_voted_slot().unwrap();
+        if slot_history.check(last_voted_slot) == Check::TooOld {
+            // We could try hard to anchor with other older votes, but opt to simplify the
+            // following logic
+            return Err(TowerError::TooOldTower(
+                last_voted_slot,
+                slot_history.oldest(),
+            ));
+        }
+
+        self.do_adjust_lockouts_after_replay(replayed_root_slot, slot_history)
+    }
+
+    fn do_adjust_lockouts_after_replay(
+        mut self,
+        replayed_root_slot: Slot,
+        slot_history: &SlotHistory,
+    ) -> Result<Self> {
+        // retained slots will be consisted only from divergent slots
+        let mut retain_flags_for_each_vote_in_reverse: Vec<_> =
+            Vec::with_capacity(self.lockouts.votes.len());
+
+        let mut still_in_future = true;
+        let mut past_outside_history = false;
+        let mut checked_slot = None;
+        let mut anchored_slot = None;
+
+        let mut root_checked = false;
+        let mut checked_slots = if let Some(root) = self.lockouts.root_slot {
+            root_checked = true;
+            vec![root]
+        } else {
+            vec![]
+        };
+        checked_slots.extend(self.voted_slots());
+
+        // iterate over votes in the newest => oldest order
+        // bail out early if bad condition is found
+        for voted_slot in checked_slots.iter().rev() {
+            checked_slot = Some(*voted_slot);
+            let check = slot_history.check(checked_slot.unwrap());
+
+            if anchored_slot.is_none() && check == Check::Found {
+                anchored_slot = checked_slot;
+            } else if anchored_slot.is_some() && check == Check::NotFound {
```

Comment:
So `slot_history` will only contain the ancestors of the snapshot bank right? So if the snapshot is of bank `3`, and our tower has votes `0, 2, 4, 5`, like so:

```
                                         0 (saved tower root)
                                          |
                                          2 (saved tower vote)
                                      /         \
                       (snapshot root) 3          4 (saved tower vote)
                                                                 \
                                                                 5 (saved tower vote)
```

The `slot_history` of bank `3` will contain slots `2` and `0`. The `anchored_slot` will be slot `2`, but slot `5` will trigger the `NotFound`.

In this case because slot 2 has a lockout of `2^3`, we will have to wait until at least some slot greater than `2 + 2^3 = 10` to vote.

Reply:
Well, I think you're missing that this iteration goes **newer slots => older slots (reverse chronological order)**: https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR807

Also, I think I've added correct test to cover the situation you just graphed there: https://github.com/solana-labs/solana/pull/10718/files#r460669570

Reply:
ah yeah you're totally right, I completely missed the `rev()`. Test looks great too, thanks!